framework/auth/tasks.py

```diff
@@ -27,3 +38,122 @@ def update_user_from_activity(user_id, login_time, cas_login=False, updates=None
         should_save = True
     if should_save:
         user.save()
+
+
+@celery_app.task()
+def update_affiliation_for_orcid_sso_users(user_id, orcid_id):
+    """This is an asynchronous task that runs during CONFIRMED ORCiD SSO logins and makes eligible
+    institution affiliations.
+    """
+    from osf.models import OSFUser
+    user = OSFUser.load(user_id)
+    if not user or not verify_user_orcid_id(user, orcid_id):
+        # This should not happen as long as this task is called at the right place at the right time.
+        error_message = f'Invalid ORCiD ID [{orcid_id}] for [{user_id}]' if user else f'User [{user_id}] Not Found'
+        logger.error(error_message)
+        sentry.log_message(error_message)
+        return
+    institution = check_institution_affiliation(orcid_id)
+    if institution:
+        logger.info(f'Eligible institution affiliation has been found for ORCiD SSO user: '
+                    f'institution=[{institution._id}], user=[{user_id}], orcid_id=[{orcid_id}]')
+        if not user.is_affiliated_with_institution(institution):
+            user.affiliated_institutions.add(institution)
+
+
+def verify_user_orcid_id(user, orcid_id):
+    """Verify that the given ORCiD ID is verified for the given user.
+    """
+    provider = EXTERNAL_IDENTITY_PROFILE.get('OrcidProfile')
+    status = user.external_identity.get(provider, {}).get(orcid_id, None)
+    return status == 'VERIFIED'
+
+
+def check_institution_affiliation(orcid_id):
+    """Check user's public ORCiD record and return eligible institution affiliations.
+
+    Note: Current implementation only support one affiliation (i.e. loop returns once eligible
+          affiliation is found, which improves performance). In the future, if we have multiple
+          institutions using this feature, we can update the loop easily.
+    """
+    from osf.models import Institution
+    from osf.models.institution import IntegrationType
+    employment_source_list = get_orcid_employment_sources(orcid_id)
+    education_source_list = get_orcid_education_sources(orcid_id)
+    via_orcid_institutions = Institution.objects.filter(
+        delegation_protocol=IntegrationType.AFFILIATION_VIA_ORCID.value,
+        is_deleted=False
+    )
+    # Check both employment and education records
+    for source in employment_source_list + education_source_list:
```

Comment:
uber nit: this will have to construct the new list and iterate through it.  Not really an issue here since these lists aren't likely to be long, but for future reference, itertools.chain saves you that extra pass by just `yielding` the items from each list in turn

Reply:
Fixed up on squash: https://github.com/CenterForOpenScience/osf.io/commit/a46390e4c783cd27bfbfb9b35c992dccf9ead1f2#diff-7a69bc6e4f832fde1a60cd28bb8fd62b26413ef033a2e575758bff9fff53bff1R89